File: packages/core/test/very-large-schema.test.ts
/**
 * What happens when the schema is very large. Two paths can blow a context window: the pruned path
 * (a normal question) and the whole-schema path (an overview question, which lists tables on
 * purpose). Both must stay bounded, and a single very wide table must not defeat the budget.
 */
import { describe, expect, it } from 'vitest';
import { estimateTokens, formatCatalogForPrompt, pruneCatalog } from '../src/catalog.js';
import type { SchemaCatalog, TableInfo } from '../src/types.js';

const BUDGET = 6000;

function tableOf(i: number, columns = 12): TableInfo {
  return {
    name: `table_${i}`,
    schema: 'app',
    kind: 'table',
    columns: Array.from({ length: columns }, (_, c) => ({
      name: c === 0 ? 'id' : `column_number_${c}`,
      dbType: 'character varying(255)',
      nullable: c !== 0,
    })),
    primaryKey: ['id'],
    foreignKeys: i > 0 ? [{ columns: ['id'], refTable: `table_${i - 1}`, refColumns: ['id'] }] : [],
    uniques: [],
    checks: [],
    indexes: [],
    source: 'db',
  };
}

function catalogOf(tables: TableInfo[]): SchemaCatalog {
  return {
    engine: 'postgres',
    schemas: ['app'],
    tables,
    enums: [],
    sequences: [],
    triggers: [],
    routines: [],
    warnings: [],
    fetchedAt: 'now',
  };
}

describe('a very large schema stays within budget', () => {
  for (const count of [200, 1000, 5000]) {
    it(`${count} tables prunes under the token budget`, () => {
      const catalog = catalogOf(Array.from({ length: count }, (_, i) => tableOf(i)));
      const pruned = pruneCatalog(catalog, 'how many rows are in table_7 by column_number_3', {
        maxSchemaTokens: BUDGET,
      });
      expect(estimateTokens(pruned.schemaText)).toBeLessThanOrEqual(BUDGET);
      // The table the question names must survive the cut.
      expect(pruned.catalog.tables.map((t) => t.name)).toContain('table_7');
    });
  }

  it('a question matching nothing still yields a usable, bounded prompt', () => {
    const catalog = catalogOf(Array.from({ length: 1000 }, (_, i) => tableOf(i)));
    const pruned = pruneCatalog(catalog, 'zzzz nothing matches this at all', { maxSchemaTokens: BUDGET });
    expect(pruned.catalog.tables.length).toBeGreaterThan(0);
    expect(estimateTokens(pruned.schemaText)).toBeLessThanOrEqual(BUDGET);
  });
});

describe('one very wide table cannot defeat the budget', () => {
  const wide = catalogOf([
    { ...tableOf(0, 800), foreignKeys: [{ columns: ['column_number_5'], refTable: 'other', refColumns: ['id'] }] },
  ]);

  it('trims columns rather than giving up', () => {
    const pruned = pruneCatalog(wide, 'count rows', { maxSchemaTokens: BUDGET });
    expect(estimateTokens(formatCatalogForPrompt(pruned.catalog))).toBeLessThanOrEqual(BUDGET);
    expect(pruned.catalog.tables[0]!.columns.length).toBeLessThan(800);
  });

  it('keeps the keys, because they carry the joins', () => {
    const pruned = pruneCatalog(wide, 'count rows', { maxSchemaTokens: BUDGET });
    const names = pruned.catalog.tables[0]!.columns.map((c) => c.name);
    expect(names).toContain('id');
    expect(names).toContain('column_number_5');
  });

  it('keeps a column the question names', () => {
    const pruned = pruneCatalog(wide, 'what is in column_number_700', { maxSchemaTokens: BUDGET });
    expect(pruned.catalog.tables[0]!.columns.map((c) => c.name)).toContain('column_number_700');
  });

  it('says how many columns are not shown, rather than hiding them silently', () => {
    const pruned = pruneCatalog(wide, 'count rows', { maxSchemaTokens: BUDGET });
    expect(pruned.catalog.tables[0]!.comment).toMatch(/of 800 columns not shown/);
    expect(pruned.schemaText).toMatch(/columns not shown/);
  });

  it('leaves a normal table alone', () => {
    const normal = catalogOf([tableOf(0, 10)]);
    const pruned = pruneCatalog(normal, 'count rows', { maxSchemaTokens: BUDGET });
    expect(pruned.catalog.tables[0]!.columns.length).toBe(10);
    expect(pruned.catalog.tables[0]!.comment).toBeUndefined();
  });

  /**
   * The trim loop and the estimator have to charge the same per column. Charging less in the loop
   * kept a whole 300-column table of sampled values and rendered 147,951 tokens against a 6,000
   * budget, while reporting strategy 'budget-trim' as if it had worked.
   */
  it('a wide table of sampled values still lands inside the budget', () => {
    const columns = Array.from({ length: 300 }, (_, i) => ({
      name: `col${i}`,
      dbType: 'text',
      nullable: true,
      comment: 'a column comment of roughly the length a real data dictionary carries for each field',
      sampledValues: Array.from({ length: 24 }, () => 'v'.repeat(80)),
    }));
    const catalog = {
      engine: 'postgres',
      schemas: [],
      tables: [
        {
          name: 'wide',
          schema: null,
          kind: 'table',
          columns,
          primaryKey: [],
          foreignKeys: [],
          indexes: [],
          comment: null,
        },
      ],
      enums: [],
      sequences: [],
      triggers: [],
      routines: [],
      warnings: [],
      fetchedAt: new Date(0).toISOString(),
    } as unknown as SchemaCatalog;

    const untrimmed = estimateTokens(formatCatalogForPrompt(catalog));
    const pruned = pruneCatalog(catalog, 'how many rows are there', { maxSchemaTokens: 6000 });
    const rendered = estimateTokens(pruned.schemaText);
    // MIN_COLUMNS_KEPT can carry a floor above the budget on purpose - an unusable stub is worse
    // than a large prompt - so the invariant is that trimming actually happened and bounded it.
    expect(rendered).toBeLessThan(untrimmed / 10);
    expect(rendered).toBeLessThan(6000 * 1.5);
    expect(pruned.catalog.tables[0]!.columns.length).toBeLessThan(300);
    expect(pruned.schemaText).toMatch(/columns not shown/);
    expect(pruned.strategy).toBe('budget-trim');
  });
});

File: packages/jetbrains/src/test/kotlin/com/rahulmahadik/asksql/ide/engine/VeryLargeSchemaTest.kt
package com.rahulmahadik.asksql.ide.engine

import com.rahulmahadik.asksql.ide.model.ColumnInfo
import com.rahulmahadik.asksql.ide.model.EngineKind
import com.rahulmahadik.asksql.ide.model.ForeignKeyInfo
import com.rahulmahadik.asksql.ide.model.SchemaCatalog
import com.rahulmahadik.asksql.ide.model.TableInfo
import com.rahulmahadik.asksql.ide.model.TableKind
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The Kotlin half of core's `very-large-schema.test.ts`. A large database must still produce a
 * prompt a model can read, and one very wide table must not defeat the budget on its own.
 */
class VeryLargeSchemaTest {

    private val budget = 6000

    private fun table(i: Int, columns: Int = 12, fks: List<ForeignKeyInfo> = emptyList()) = TableInfo(
        schema = "app",
        name = "table_$i",
        kind = TableKind.TABLE,
        columns = (0 until columns).map {
            ColumnInfo(name = if (it == 0) "id" else "column_number_$it", dbType = "character varying(255)", nullable = it != 0)
        },
        primaryKey = listOf("id"),
        foreignKeys = fks,
    )

    private fun catalog(tables: List<TableInfo>) =
        SchemaCatalog(engine = EngineKind.POSTGRES, schemas = listOf("app"), tables = tables)

    private fun settings() = CatalogPruner.PrunerSettings(maxSchemaTokens = budget)

    private fun tokens(text: String) = kotlin.math.ceil(text.length / 4.0).toInt()

    @Test
    fun `a large schema prunes under the token budget`() {
        for (count in listOf(200, 1000, 5000)) {
            val cat = catalog((0 until count).map { table(it) })
            val pruned = CatalogPruner.pruneCatalog(cat, "how many rows are in table_7 by column_number_3", settings())
            assertTrue("$count tables: ${tokens(pruned.schemaText)} tokens", tokens(pruned.schemaText) <= budget)
            assertTrue("table_7 must survive", pruned.catalog.tables.any { it.name == "table_7" })
        }
    }

    @Test
    fun `a question matching nothing still yields a bounded prompt`() {
        val cat = catalog((0 until 1000).map { table(it) })
        val pruned = CatalogPruner.pruneCatalog(cat, "zzzz nothing matches this at all", settings())
        assertTrue(pruned.catalog.tables.isNotEmpty())
        assertTrue(tokens(pruned.schemaText) <= budget)
    }

    private fun wide() = catalog(
        listOf(table(0, columns = 800, fks = listOf(ForeignKeyInfo(columns = listOf("column_number_5"), refTable = "other", refColumns = listOf("id"))))),
    )

    @Test
    fun `one very wide table is trimmed rather than giving up`() {
        val pruned = CatalogPruner.pruneCatalog(wide(), "count rows", settings())
        assertTrue(tokens(pruned.schemaText) <= budget)
        assertTrue(pruned.catalog.tables[0].columns.size < 800)
    }

    /** Keys carry the joins, so they survive the cut whatever else does not. */
    @Test
    fun `trimming keeps the keys and a column the question names`() {
        val kept = CatalogPruner.pruneCatalog(wide(), "count rows", settings()).catalog.tables[0].columns.map { it.name }
        assertTrue(kept.contains("id"))
        assertTrue(kept.contains("column_number_5"))
        val named = CatalogPruner.pruneCatalog(wide(), "what is in column_number_700", settings()).catalog.tables[0].columns.map { it.name }
        assertTrue(named.contains("column_number_700"))
    }

    @Test
    fun `trimming says how many columns are not shown`() {
        val pruned = CatalogPruner.pruneCatalog(wide(), "count rows", settings())
        assertTrue(pruned.catalog.tables[0].comment!!.contains("of 800 columns not shown"))
        assertTrue(pruned.schemaText.contains("columns not shown"))
    }

    @Test
    fun `a normal table is left alone`() {
        val pruned = CatalogPruner.pruneCatalog(catalog(listOf(table(0, columns = 10))), "count rows", settings())
        assertTrue(pruned.catalog.tables[0].columns.size == 10)
        assertTrue(pruned.catalog.tables[0].comment == null)
    }
}
